proposals/0194-deprecate-rent-exemption-threshold.md

```diff
@@ -0,0 +1,123 @@
+---
+simd: '0194'
+title: Deprecate Rent Exemption Threshold
+authors:
+  - Dean Little (@deanmlittle)
+  - Leonardo Donatacci (@L0STE)
+  - febo (Anza)
+category: Standard/Meta
+type: Core
+status: Draft
+created: 2024-11-13
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This proposal aims to eliminate the use of floating-point operations to 
+determine whether an account is rent exempt or not in programs by deprecating
+the use of the `exempt_threshold` (`f64`) value. 
+
+More specifically, rename `lamports_per_byte_year` to `lamports_per_byte`, 
+change default value from `3480` to `6960`, change `exemption_threshold` to 
+`1.0` and deprecate it from the protocol. This will enable us to remove `f64` 
+math from all Rent-related SDKs and onchain programs moving forward.
+
+## Motivation
+
+Emulating floating point math is expensive inside of SVM due to the lack of 
+native floating point number support. This makes calculating the rent exempt 
+threshold cost `~248` more CUs than it would if we were to simply use a `u64`. 
+This is due to the `exemption_threshold` (`f64`), which is currently set to 
+`2.0` years. Since rent exemption is no longer time-based, we have the 
+opportunity to make this commonly used calculation much more performant and 
+simplify our tooling without affecting the existing API. It also simplifies any 
+further changes we may want to make to Rent down the line.
+
+## New Terminology
+
+`lamports_per_byte` - the number of lamports required to pay for 1 byte of 
+account storage.
+
+## Detailed Design
+
+Set the value of `DEFAULT_EXEMPTION_THRESHOLD` from `2.0` to `1.0`, and 
+deprecate it from the protocol.
+
+```rs
+pub const DEFAULT_EXEMPTION_THRESHOLD: f64 = 1.0;
+```
+
+Set `DEFAULT_LAMPORTS_PER_BYTE` from its current value of `3480` (the `u64` 
+value of `1_000_000_000 / 100 * 365 / (1024 * 1024)`), to `6960`; double its 
+current `u64` value, to counteract reducing the exemption threshold above.
+
+```rs
+pub const DEFAULT_LAMPORTS_PER_BYTE: u64 = 6960;
+```
+
+Rename `lamports_per_byte_year` in the Rent account to `lamports_per_byte` to 
+reflect that Rent is no longer time-based and officially deprecate 
+`exemption_threshold`.
+
+```rs
+pub struct Rent {
+    pub lamports_per_byte: u64,
+    #[deprecated(since = "2.X.X", note = "Use only `lamports_per_byte` 
+instead")]
+    pub exemption_threshold: f64,
+    pub burn_percent: u8,
+}
+```
+
+ Remove all `f64` math from Rent calculations in all SDKs and onchain programs 
+moving forwards, replacing it with simple `u64` math, ie:
+
+```rs
+/// Minimum balance due for rent-exemption of a given account data size.
+pub fn minimum_balance(&self, data_len: usize) -> u64 {
+    (ACCOUNT_STORAGE_OVERHEAD + data_len as u64) * self.lamports_per_byte
+}
```

Comment:
Making this calculation faster is great. I'm just wondering if renaming `lamports_per_byte_year` to `lamports_per_byte` might be confusing for people grepping for `lamports_per_byte_year`. As rent is going away anyway I wonder if this would add more short-term confusion?

If the goal of this SIMD is to remove the floating point operations from this calculation, I wonder if it would be simpler to mark `exemption_threshold` as deprecated and make the following change in the SDK to remove the floating-point calculation:

```suggestion
    /// Minimum balance due for rent-exemption of a given account data size.
    pub fn minimum_balance(&self, data_len: usize) -> u64 {
        let bytes = data_len as u64;
        // exemption_threshold has been deprecated and is always 2
        let exemption_threshold = 2;
        ((ACCOUNT_STORAGE_OVERHEAD + bytes) * self.lamports_per_byte_year) * exemption_threshold
    }
```

Thoughts? That said, I am ok with the proposed change, just wanted to throw a simpler alternative out there.

Reply:
Yes, this was one of the discussed alternatives [above](https://github.com/solana-foundation/solana-improvement-documents/pull/194#discussion_r1853351246). If we simply decide "This number is now 2 forever, we promise not to change it", it has the least moving parts. It doesn't require a feature gate either. We can simply change all our libraries to ignore the f64 value stored in the Rent account/sysvar and substitute with our own u64. The main downsides are that we waste 1 extra CU multiplying by 2 every time unnecessarily, and instead of removing existing protocol crust, we introduce a new piece where we have this random number 2 thrown in there for no reason. I prefer changing the value to 1 to remove it from the equation, but both are fine changes imo.

Reply:
> If we simply decide "This number is now 2 forever, we promise not to change it", it has the least moving parts. It doesn't require a feature gate either. We can simply change all our libraries to ignore the f64 value stored in the Rent account/sysvar and substitute with our own u64.

I would prefer this option, but I am happy with either.

> What if we were to just hardcode the value of rent in our libraries too?

Do you mean making the minimum balance independent of the size of the account? I think that we should still take the size of the account into consideration.

Reply:
No, I mean making the lamports per bytes for rent exemption static. Right now, the reason we have a syscall and/or Rent account at all is to tell us:
1. how many `lamports_per_byte_year` we are paying, which is multiplied by
2. the `exemption_threshold` of `2.0f64` years

If these values will not change between now and when Rent is removed altogether, we could simply hardcode both values into our libraries enabling you to statically derive the correct amount of rent to pay without invoking a syscall at all.